File: backend/services/traffic_simulator.py
import random
import time
from typing import List, Dict

EDGES = [
    ("kicukiro-sonatube", 50, 1.8),
    ("sonatube-gikondo", 45, 1.5),
    ("sonatube-magerwa", 60, 2.0),
    ("gikondo-rwandex", 40, 1.3),
    ("rwandex-nyarugenge", 45, 2.4),
    ("magerwa-remera", 60, 2.3),
    ("remera-kimihurura", 50, 2.0),
    ("kimihurura-downtown", 45, 1.8),
    ("kimihurura-kacyiru", 50, 1.9),
    ("kacyiru-downtown", 55, 2.2),
    ("nyarugenge-downtown", 40, 1.2),
    ("nyabugogo-downtown", 50, 1.5),
]


def generate_traffic_data(peak_hour: bool = False) -> List[Dict]:
    now = int(time.time())
    peak_mult = 1.6 if peak_hour else 0.8
    results = []

    for eid, base_speed, dist in EDGES:
        density = max(5.0, min(100.0, random.gauss(35 * peak_mult, 15)))
        speed = max(8.0, base_speed * (1 - density / 110) + random.gauss(0, 3))
        vehicles = int(density * dist * 0.9)
        flow = int(density * speed * 0.8)

        results.append({
            "edgeId": eid,
            "speed": round(speed, 1),
            "vehicleCount": vehicles,
            "density": round(density, 1),
            "flowRate": flow,
            "timestamp": now,
        })

    return results

File: src/lib/trafficSimulator.ts
// Simulated GPS traffic data for each road segment
// Generates realistic speed / vehicle density values and updates them over time

import { EDGES } from './roadNetwork';

export interface TrafficData {
  edgeId: string;
  speed: number;         // current avg speed km/h
  vehicleCount: number;  // vehicles currently on segment
  density: number;       // 0-100 relative density
  flowRate: number;      // vehicles/hour
  timestamp: number;
}

export interface SimulatorOptions {
  peakHour: boolean;
}

// Deterministic pseudo-random per-edge fluctuation so markers don't jitter wildly
function noise(seed: number) {
  const x = Math.sin(seed * 12.9898) * 43758.5453;
  return x - Math.floor(x);
}

export function generateTrafficData(opts: SimulatorOptions = { peakHour: false }): TrafficData[] {
  const now = Date.now();
  const tick = Math.floor(now / 3000); // changes every 3s

  return EDGES.map((edge, i) => {
    // Base congestion factor per edge (some roads are always busier)
    const edgeBias = noise(i + 1);
    const timeNoise = noise(tick + i * 7);

    // Peak hour adds heavy congestion on main corridors
    const peakMultiplier = opts.peakHour ? 1.6 + edgeBias * 0.6 : 0.7 + edgeBias * 0.5;

    // Vehicle density 0-100
    const density = Math.min(100, Math.max(5, 25 * peakMultiplier + timeNoise * 30));

    // Speed decreases as density increases (Greenshields-like model)
    const speedFactor = Math.max(0.15, 1 - density / 110);
    const speed = Math.max(8, edge.baseSpeed * speedFactor + (timeNoise - 0.5) * 6);

    // Flow rate (vehicles/hour) = density * speed / distance normalization
    const flowRate = Math.round(density * speed * 0.8);

    // Vehicle count approximation for segment
    const vehicleCount = Math.round(density * edge.distance * 0.9);

    return {
      edgeId: edge.id,
      speed: Math.round(speed * 10) / 10,
      vehicleCount,
      density: Math.round(density),
      flowRate,
      timestamp: now,
    };
  });
}
